[event based orderbook] Implement StableXOrderBookReading for State

e00E opened the pull request:
I am going to add tests when this is fully implemented.

e00E committed:
Implement StableXOrderBookReading for State

fleupold commented on driver/src/orderbook/streamed/orderbook.rs:
```diff
@@ -0,0 +1,85 @@
+    fn get_auction_data(&self, _index: U256) -> Result<(AccountState, Vec<Order>)> {
+        // TODO: Handle future batch ids for when we want to do optimistic solving.
+        let state = self.create_state()?;
+        let (account_state, orders) = state.orderbook(Batch::Current)?;
```
I think here it would be handy if Future(index) would also be valid if it's the current batch id.

fleupold commented on driver/src/orderbook/streamed/orderbook.rs:
```diff
@@ -0,0 +1,85 @@
+/// The key by which events are sorted.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct Key {
+    block_number: u64,
```
nit: Is block_number really needed? Shoudn't hash and index be uniquely identifying an event?

e00E replied:
Not needed for identifying but needed for ordering.

fleupold commented on driver/src/orderbook/streamed/orderbook.rs:
```diff
@@ -0,0 +1,85 @@
+
+/// The key by which events are sorted.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct Key {
```
Not important, but if this was called `EventSortKey` or the like we wouldn't need the comment above. I think generally it's nice to replace comments with more explicit naming (as comments tend to not get updated when the code/logic changes).

fleupold reviewed (approved):
Nice, we are almost there!

fleupold commented on driver/src/orderbook/streamed/orderbook.rs:
```diff
@@ -0,0 +1,85 @@
+    batch_id: BatchId,
+}
+
+pub struct Orderbook {
```
Would it make sense to give this struct a prefixed name such as `EventBasedOrderbook`?

e00E replied:
I don't think so because it is already in the streamed module.

nlordell replied:
In general I also tend to prefer shorter names if the module clearly states what its for. Outside of this module we can use `streamed::Orderbook` so its clear that its a streamed version of the orderbook.

e00E committed:
rename Key

nlordell commented on driver/src/orderbook/streamed/orderbook.rs:
```diff
@@ -0,0 +1,84 @@
+    block_number: u64,
+    /// Is included to differentiate events from the same block number but different blocks which
+    /// can happen during reorgs.
+    block_hash: H256,
```
Should `log_index` come before `block_hash` in terms of ordering?

e00E replied:
It does not matter. The only time during which there can be multiple blocks of the same number is during reorg before all old events have been deleted. And in that case the resulting state isn't correct anyway.

nlordell replied:
Actually in that case, having `block_hash` come first as it is makes a little more sense, as the events will be grouped together.

nlordell commented on driver/src/orderbook/streamed/orderbook.rs:
```diff
@@ -0,0 +1,84 @@
+        block_number: u64,
+        log_index: usize,
+        block_hash: H256,
+        block_timestamp: u64,
```
This might be problematic as querying the node for all past time stamps can be really slow.

e00E replied:
I don't see a way around it. I have tested this on rinkeby by querying the node for all timestamps needed and caching the last timestamp to speed up multiple events from the same block.

nlordell replied:
The only way I see would be to handle past events slightly differently than future events, so for past events you apply all the events under the current batch ID (queried from the smart contract for example) and then for incoming events you can query the block timestamp for each (since they will be way more sparse).

Definitely a separate issue/PR though.

e00E committed:
fix rebase

e00E committed:
Implement updating of orderbook with events (#761)

This implements the task that updates the orderbook with incoming events
from a stream. After this we still need to actually create the task with
the event stream from the contract which will happen in the next PR.
